# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
import pandas as pd
import numpy as np



# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. You can find the data in Ironhack's database (db: weather, table: austin_weather). 

In [2]:
austin = pd.read_csv('../data/austin_weather.csv')


#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
austin.dtypes

Date                          object
TempHighF                      int64
TempAvgF                       int64
TempLowF                       int64
DewPointHighF                 object
DewPointAvgF                  object
DewPointLowF                  object
HumidityHighPercent           object
HumidityAvgPercent            object
HumidityLowPercent            object
SeaLevelPressureHighInches    object
SeaLevelPressureAvgInches     object
SeaLevelPressureLowInches     object
VisibilityHighMiles           object
VisibilityAvgMiles            object
VisibilityLowMiles            object
WindHighMPH                   object
WindAvgMPH                    object
WindGustMPH                   object
PrecipitationSumInches        object
Events                        object
dtype: object

In [4]:
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

In [5]:
austin.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [6]:
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [7]:
austin['DewPointAvgF'].unique()


array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [8]:
#I noticed that one of the values is a '-' which is why pandas treated this variable as an object instead of int64.


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [9]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [10]:
austin['DewPointHighF'] = pd.to_numeric(austin['DewPointHighF'], errors='coerce')
#austin['Date'] = pd.to_datetime(austin['Date'])
cols = austin.columns.drop(['Date','Events'])
#cols = austin.columns[austin.dtypes.eq('object')]
austin[cols] = austin[cols].apply(pd.to_numeric, errors='coerce')
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

In [11]:
austin.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0  ...                      29.68                      29.59   
1  ...                      30.13                      29.87   
2  ...                      30.49                      30.41   
3  ...                      30.45                      30.30   
4  ...                      30.33                      30.27   

   VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
0                 10.0                 7.0                 2.0         20.0   
1                 10.0                10.0                 5.0         16.0   
2                 10.0                10.0                10.0          8.0   
3                 10.0                10.0                 7.0         12.0   
4                 10.0                10.0                 7.0         10.0   

   WindAvgMPH  WindGustMPH  PrecipitationSumInches               Events  
0         4.0         31.0                    0.46  Rain , Thunderstorm  
1         6.0         25.0                    0.00                       
2         3.0         12.0                    0.00                       
3         4.0         20.0                    0.00                       
4         2.0         16.0                     NaN                       

[5 rows x 21 columns]

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [12]:
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [13]:
austin['missing_values'] = austin.isnull().any(axis=1)
#missing_values = austin.isnull().any(axis=1).sum()
austin.head(60)


Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0   2013-12-21         74        60        45           67.0          49.0   
1   2013-12-22         56        48        39           43.0          36.0   
2   2013-12-23         58        45        32           31.0          27.0   
3   2013-12-24         61        46        31           36.0          28.0   
4   2013-12-25         58        50        41           44.0          40.0   
5   2013-12-26         57        48        39           39.0          36.0   
6   2013-12-27         60        53        45           41.0          39.0   
7   2013-12-28         62        51        40           43.0          39.0   
8   2013-12-29         64        50        36           49.0          41.0   
9   2013-12-30         44        40        35           31.0          26.0   
10  2013-12-31         55        46        36           31.0          28.0   
11  2014-01-01         69        54        39           51.0          42.0   
12  2014-01-02         55        44        33           39.0          26.0   
13  2014-01-03         58        43        28           37.0          22.0   
14  2014-01-04         71        57        42           55.0          48.0   
15  2014-01-05         59        47        34           54.0          32.0   
16  2014-01-06         36        29        22           15.0           8.0   
17  2014-01-07         48        35        22           29.0          11.0   
18  2014-01-08         53        47        40           51.0          45.0   
19  2014-01-09         70        62        53           60.0          55.0   
20  2014-01-10         72        65        57           64.0          61.0   
21  2014-01-11         75        62        48           54.0          37.0   
22  2014-01-12         67        57        46           58.0          47.0   
23  2014-01-13         71        57        43           60.0          36.0   
24  2014-01-14         76        57        37           31.0          25.0   
25  2014-01-15         64        52        40           29.0          23.0   
26  2014-01-16         72        56        40           31.0          27.0   
27  2014-01-17         64        54        43           31.0          20.0   
28  2014-01-18         70        54        37           41.0          33.0   
29  2014-01-19         72        59        45           41.0          30.0   
30  2014-01-20         82        64        46           51.0          36.0   
31  2014-01-21         63        51        38           28.0          23.0   
32  2014-01-22         70        54        37           39.0          27.0   
33  2014-01-23         56        42        27           42.0          29.0   
34  2014-01-24         40        33        26           22.0          17.0   
35  2014-01-25         70        48        26           38.0          27.0   
36  2014-01-26         76        60        43           45.0          39.0   
37  2014-01-27         60        47        34           43.0          23.0   
38  2014-01-28         39        33        27           19.0          14.0   
39  2014-01-29         49        38        27           19.0          13.0   
40  2014-01-30         65        48        31           51.0          36.0   
41  2014-01-31         80        68        55           59.0          54.0   
42  2014-02-01         76        66        55           62.0          59.0   
43  2014-02-02         55        45        35           40.0          33.0   
44  2014-02-03         46        40        34           38.0          33.0   
45  2014-02-04         64        53        42           45.0          41.0   
46  2014-02-05         48        40        31           42.0          26.0   
47  2014-02-06         32        29        26           17.0          15.0   
48  2014-02-07         38        34        29           29.0          24.0   
49  2014-02-08         65        51        37           42.0          36.0   
50  2014-02-09         76       

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [14]:
print(austin.shape)
count = austin[austin['missing_values']==True]
count.shape

(1319, 22)


(136, 22)

#### Calculate the ratio of missing rows to total rows

In [15]:
ratio = len(count.index)/len(austin.index)
ratio


0.10310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [16]:
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
missing_values                  0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [17]:
precipsuminches = 124/len(austin.index)
precipsuminches

0.09401061410159212

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [18]:
austin.drop(['PrecipitationSumInches'], axis=1, inplace=True)

# Print `austin` to confirm the column is indeed removed
austin

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
0                   57.0  ...                      29.68   
1                   43.0  ...                      30.13   
2                   27.0  ...                      30.49   
3                   22.0  ...                      30.45   
4                   56.0  ...                      30.33   
...                  ...  ...                        ...   
1314                25.0  ...                      29.97   
1315                20.0  ...                      29.90   
1316                19.0  ...                      29.86   
1317                27.0  ...                      29.91   
1318                22.0  ...                      29.97   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                         29.59                 10.0                 7.0   
1                         29.87                 10.0                10.0   
2                         30.41                 10.0                10.0   
3                         30.30                 10.0                10.0   
4                         30.27                 10.0                10.0   
...                         ...                  ...                 ...   
1314                      29.88                 10.0                10.0   
1315                      29.81                 10.0                10.0   
1316                      29.79                 10.0                10.0   
1317                      29.87                 10.0                10.0   
1318                      29.91                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                    2.0         20.0         4.0         31.0   
1                    5.0         16.0         6.0         25.0   
2                   10.0          8.0         3.0         12.0   
3                    7.0         12.0         4.0         20.0   
4                    7.0         10.0         2.0         16.0   
...                  ...          ...         ...          ...   
1314                10.0         12.0         5.0         21.0   
1315                10.0         14.0         5.0         20.0   
1316                10.0         12.0         4.0         17.0   


#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [19]:
austin_fixed = austin.interpolate(inplace=False)
austin_fixed
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
missing_values                0
dtype: int64

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [20]:
austin.isna().sum()

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
missing_values                 0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [21]:
austin_fixed['Events'].unique()
austin_fixed['Events'].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [22]:
#the largest number of events is rain at 192

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [23]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
austin_fixed['Snow'] = 0
austin_fixed['Fog'] = 0
austin_fixed['Rain'] = 0
austin_fixed['Thunderstorm'] = 0

# Print your new dataframe to check whether new columns have been created:
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                   57.0  ...                 2.0         20.0         4.0   
1                   43.0  ...                 5.0         16.0         6.0   
2                   27.0  ...                10.0          8.0         3.0   
3                   22.0  ...                 7.0         12.0         4.0   
4                   56.0  ...                 7.0         10.0         2.0   
...                  ...  ...                 ...          ...         ...   
1314                25.0  ...                10.0         12.0         5.0   
1315                20.0  ...                10.0         14.0         5.0   
1316                19.0  ...                10.0         12.0         4.0   
1317                27.0  ...                10.0         13.0         4.0   
1318                22.0  ...                10.0         12.0         4.0   

      WindGustMPH               Events  missing_values  Snow  Fog  Rain  \
0            31.0  Rain , Thunderstorm           False     0    0     0   
1            25.0                                False     0    0     0   
2            12.0                                False     0    0     0   
3            20.0                                False     0    0     0   
4            16.0                                 True     0    0     0   
...           ...                  ...             ...   ...  ...   ...   
1314         21.0                                False     0    0     0   
1315         20.0                                False     0    0     0   
1316         17.0                                False     0    0     0   
1317         20.0                                False     0    0     0   
1318         20.0                                False     0    0     0   

     Thunderstorm  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
1314            0  
1315            0  
1316            0  
1317            0  
1318            0  

[1319 rows x 25 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [24]:
import re
#austin_fixed.Series.str.contains('Rain', regex=False)
austin_fixed['Rain'] = austin_fixed['Events'].str.contains('Rain').astype(int)
austin_fixed['Snow'] = austin_fixed['Events'].str.contains('Snow').astype(int)
austin_fixed['Fog'] = austin_fixed['Events'].str.contains('Fog').astype(int)
austin_fixed['Thunderstorm'] = austin_fixed['Events'].str.contains('Thunderstorm').astype(int)
austin_fixed['Rain'].value_counts()

0    942
1    377
Name: Rain, dtype: int64

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [25]:
austin_fixed.head(60)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0   2013-12-21         74        60        45           67.0          49.0   
1   2013-12-22         56        48        39           43.0          36.0   
2   2013-12-23         58        45        32           31.0          27.0   
3   2013-12-24         61        46        31           36.0          28.0   
4   2013-12-25         58        50        41           44.0          40.0   
5   2013-12-26         57        48        39           39.0          36.0   
6   2013-12-27         60        53        45           41.0          39.0   
7   2013-12-28         62        51        40           43.0          39.0   
8   2013-12-29         64        50        36           49.0          41.0   
9   2013-12-30         44        40        35           31.0          26.0   
10  2013-12-31         55        46        36           31.0          28.0   
11  2014-01-01         69        54        39           51.0          42.0   
12  2014-01-02         55        44        33           39.0          26.0   
13  2014-01-03         58        43        28           37.0          22.0   
14  2014-01-04         71        57        42           55.0          48.0   
15  2014-01-05         59        47        34           54.0          32.0   
16  2014-01-06         36        29        22           15.0           8.0   
17  2014-01-07         48        35        22           29.0          11.0   
18  2014-01-08         53        47        40           51.0          45.0   
19  2014-01-09         70        62        53           60.0          55.0   
20  2014-01-10         72        65        57           64.0          61.0   
21  2014-01-11         75        62        48           54.0          37.0   
22  2014-01-12         67        57        46           58.0          47.0   
23  2014-01-13         71        57        43           60.0          36.0   
24  2014-01-14         76        57        37           31.0          25.0   
25  2014-01-15         64        52        40           29.0          23.0   
26  2014-01-16         72        56        40           31.0          27.0   
27  2014-01-17         64        54        43           31.0          20.0   
28  2014-01-18         70        54        37           41.0          33.0   
29  2014-01-19         72        59        45           41.0          30.0   
30  2014-01-20         82        64        46           51.0          36.0   
31  2014-01-21         63        51        38           28.0          23.0   
32  2014-01-22         70        54        37           39.0          27.0   
33  2014-01-23         56        42        27           42.0          29.0   
34  2014-01-24         40        33        26           22.0          17.0   
35  2014-01-25         70        48        26           38.0          27.0   
36  2014-01-26         76        60        43           45.0          39.0   
37  2014-01-27         60        47        34           43.0          23.0   
38  2014-01-28         39        33        27           19.0          14.0   
39  2014-01-29         49        38        27           19.0          13.0   
40  2014-01-30         65        48        31           51.0          36.0   
41  2014-01-31         80        68        55           59.0          54.0   
42  2014-02-01         76        66        55           62.0          59.0   
43  2014-02-02         55        45        35           40.0          33.0   
44  2014-02-03         46        40        34           38.0          33.0   
45  2014-02-04         64        53        42           45.0          41.0   
46  2014-02-05         48        40        31           42.0          26.0   
47  2014-02-06         32        29        26           17.0          15.0   
48  2014-02-07         38        34        29           29.0          24.0   
49  2014-02-08         65        51        37           42.0          36.0   
50  2014-02-09         76       

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [26]:
austin_fixed.drop(['Events'], axis=1, inplace=True)
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  missing_values  Snow  Fog  Rain  \
0            20.0         4.0         31.0           False     0    0     1   
1            16.0         6.0         25.0           False     0    0     0   
2             8.0         3.0         12.0           False     0    0     0   
3            12.0         4.0         20.0           False     0    0     0   
4            10.0         2.0         16.0            True     0    0     0   
...           ...         ...          ...             ...   ...  ...   ...   
1314         12.0         5.0         21.0           False     0    0     0   
1315         14.0         5.0         20.0           False     0    0     0   
1316         12.0         4.0         17.0           False     0    0     0   
1317         13.0         4.0         20.0           False     0    0     0   
1318         12.0         4.0         20.0           False     0    0     0   

      Thunderstorm  
0                1  
1                0  
2                0  
3                0  
4                0  
...            ...  
1314             0  
1315             0  
1316             0  
1317             0  
1318             0  

[1319 rows x 24 columns]

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [27]:
from datetime import datetime

#my code -- I refreshed Kernel and it finally worked
austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d').toordinal())

#code received from Tiina which is the same code
#austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d').toordinal())


#### Print `austin_fixed` to check your `Date` column.

In [28]:
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     735223         74        60        45           67.0          49.0   
1     735224         56        48        39           43.0          36.0   
2     735225         58        45        32           31.0          27.0   
3     735226         61        46        31           36.0          28.0   
4     735227         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1314  736537        103        89        75           71.0          67.0   
1315  736538        105        91        76           71.0          64.0   
1316  736539        107        92        77           72.0          64.0   
1317  736540        106        93        79           70.0          68.0   
1318  736541         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  missing_values  Snow  Fog  Rain  \
0            20.0         4.0         31.0           False     0    0     1   
1            16.0         6.0         25.0           False     0    0     0   
2             8.0         3.0         12.0           False     0    0     0   
3            12.0         4.0         20.0           False     0    0     0   
4            10.0         2.0         16.0            True     0    0     0   
...           ...         ...          ...             ...   ...  ...   ...   
1314         12.0         5.0         21.0           False     0    0     0   
1315         14.0         5.0         20.0           False     0    0     0   
1316         12.0         4.0         17.0           False     0    0     0   
1317         13.0         4.0         20.0           False     0    0     0   
1318         12.0         4.0         20.0           False     0    0     0   

      Thunderstorm  
0                1  
1                0  
2                0  
3                0  
4                0  
...            ...  
1314             0  
1315             0  
1316             0  
1317             0  
1318             0  

[1319 rows x 24 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [35]:
columns= ['TempHighF', 'TempLowF', 'DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 'HumidityAvgPercent',
 'HumidityLowPercent', 'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches', 'SeaLevelPressureLowInches','VisibilityHighMiles',
 'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 'WindGustMPH', 'Snow', 'Fog', 'Rain','Thunderstorm']
austin_fixed_x = austin_fixed[columns]
austin_fixed_y = austin_fixed['TempAvgF']

#another way to write austin_fixed_y
#austin_fixed_y = austin_fixed.drop('TempAvgF', axis = 1)

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [36]:
from sklearn.model_selection import train_test_split


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [37]:
X_train, X_test, y_train, y_test = train_test_split(austin_fixed_x, austin_fixed_y, test_size=0.2)

In [38]:
X_train

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
616          95        71           65.0          63.0          60.0   
858          90        62           71.0          58.0          36.0   
106          61        52           51.0          49.0          45.0   
625         100        76           74.0          70.0          64.0   
150          89        67           65.0          62.0          59.0   
...         ...       ...            ...           ...           ...   
690          85        69           71.0          64.0          27.0   
677          81        51           64.0          54.0          49.0   
1193         86        70           68.0          66.0          60.0   
983          92        73           73.0          72.0          70.0   
787          82        43           47.0          40.0          32.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
616                  79.0                55.0                31.0   
858                  93.0                54.0                15.0   
106                  96.0                76.0                55.0   
625                  88.0                61.0                33.0   
150                  84.0                63.0                42.0   
...                   ...                 ...                 ...   
690                  93.0                62.0                31.0   
677                  96.0                66.0                35.0   
1193                 90.0                67.0                44.0   
983                  94.0                73.0                52.0   
787                  71.0                49.0                26.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  ...  \
616                        30.04                      29.98  ...   
858                        29.78                      29.69  ...   
106                        29.89                      29.81  ...   
625                        29.98                      29.91  ...   
150                        30.08                      30.01  ...   
...                          ...                        ...  ...   
690                        30.00                      29.87  ...   
677                        29.97                      29.89  ...   
1193                       29.87                      29.78  ...   
983                        30.04                      29.97  ...   
787                        30.09                      29.99  ...   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
616                  10.0                10.0                10.0   
858                  10.0                 8.0                 1.0   
106                  10.0                 7.0                 2.0   
625                  10.0                10.0                10.0   
150                  10.0                10.0                10.0   
...                   ...                 ...                 ...   
690                  10.0                10.0                10.0   
677                  10.0                10.0                10.0   
1193                 10.0                 8.0                 4.0   
983                  10.0                10.0                10.0   
787                  10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
616           9.0         3.0         15.0     0    0     0             0  
858          16.0         5.0         37.0     0    0     1             1  
106          12.0         5.0         19.0     0    0     1             1  
625          13.0         6.0         20.0     0    0     0             0  
150          14.0         9.0         26.0     0    0     0             0  
...           ...         ...          ...   ...  ...   ...           ...  
690          13.0         7.0         25.0     0    0     0             0  
677          12.0         4.0         18.0     0    0     

In [39]:
X_test

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
781          74        38           36.0          33.0          30.0   
1172         77        61           67.0          51.0          34.0   
632          90        64           68.0          57.0          51.0   
17           48        22           29.0          11.0           4.0   
844          72        58           61.0          60.0          58.0   
...         ...       ...            ...           ...           ...   
419          70        39           44.0          38.0          32.0   
54           69        32           38.0          34.0          28.0   
950          97        75           76.0          73.0          71.0   
327          41        31           17.0          13.0           9.0   
462          84        48           49.0          44.0          40.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
781                  82.0                54.0                25.0   
1172                 93.0                61.0                29.0   
632                  79.0                53.0                27.0   
17                   68.0                43.0                17.0   
844                 100.0                83.0                66.0   
...                   ...                 ...                 ...   
419                  79.0                58.0                37.0   
54                   85.0                58.0                30.0   
950                  94.0                70.0                46.0   
327                  58.0                43.0                27.0   
462                  86.0                54.0                22.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  ...  \
781                        30.41                      30.30  ...   
1172                       30.29                      30.12  ...   
632                        30.07                      30.01  ...   
17                         30.66                      30.52  ...   
844                        30.09                      30.04  ...   
...                          ...                        ...  ...   
419                        30.38                      30.31  ...   
54                         30.11                      30.02  ...   
950                        30.04                      29.98  ...   
327                        30.58                      30.52  ...   
462                        30.14                      30.08  ...   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
781                  10.0                10.0                10.0   
1172                 10.0                10.0                 5.0   
632                  10.0                10.0                10.0   
17                   10.0                10.0                10.0   
844                  10.0                 8.0                 2.0   
...                   ...                 ...                 ...   
419                  10.0                10.0                10.0   
54                   10.0                10.0                10.0   
950                  10.0                10.0                 8.0   
327                  10.0                10.0                10.0   
462                  10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
781           9.0         3.0         16.0     0    0     0             0  
1172         17.0         6.0         25.0     0    0     0             0  
632          12.0         5.0         17.0     0    0     0             0  
17           13.0         5.0         22.0     0    0     0             0  
844          10.0         4.0         20.0     0    0     1             0  
...           ...         ...          ...   ...  ...   ...           ...  
419          10.0         2.0         15.0     0    0     0             0  
54           10.0         4.0         19.0     0    0     

In [40]:
y_train

616     83
858     76
106     57
625     88
150     78
        ..
690     77
677     66
1193    78
983     83
787     63
Name: TempAvgF, Length: 1055, dtype: int64

In [41]:
y_test

781     56
1172    69
632     77
17      35
844     65
        ..
419     55
54      51
950     86
327     36
462     66
Name: TempAvgF, Length: 264, dtype: int64

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [32]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [33]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [34]:
# Your code here:
